JWBestPracticeApps/AdvancedPlayer/AdvancedPlayer/ViewController.swift

```diff
@@ -0,0 +1,299 @@
+        super.jwplayer(player, adEvent: event)
+
+
+        print("Ad Event type: \(event[.type] ?? "unknown")")
+
+        switch event.type {
+        case .impression, .meta:
+            // MARK: - 4. Detect whether an upcoming ad is skippable
+            // If the ad has a skip offset >= 0, it's skippable.
+            if let skipOffset = event[.skipOffset] as? TimeInterval, skipOffset >= 0 {
+                self.skipOffset = skipOffset
+                DispatchQueue.main.async { [weak self] in
+                    self?.customControls.skipButton.isHidden = false
+                }
+            }
+        case .adBreakStart:
+            DispatchQueue.main.async { [weak self] in
+                self?.customControls.learnMoreButton.isHidden = false
+                self?.customControls.progressBar.tintColor = .orange
+                self?.customControls.progressBar.isUserInteractionEnabled = false
+            }
+        case .adBreakEnd:
+            DispatchQueue.main.async { [weak self] in
+                self?.customControls.learnMoreButton.isHidden = true
+                self?.customControls.progressBar.tintColor = .red
+                self?.customControls.progressBar.isUserInteractionEnabled = true
+            }
+        case .complete:
+            DispatchQueue.main.async { [weak self] in
+                self?.customControls.skipButton.isHidden = true
+            }
+        case .play:
+            DispatchQueue.main.async { [weak self] in
+                self?.customControls.playPauseButton.setImage(UIImage(systemName: "pause"), for: .normal)
+            }
+        case .pause:
+            DispatchQueue.main.async { [weak self] in
+                self?.customControls.playPauseButton.setImage(UIImage(systemName: "play"), for: .normal)
+            }
+        case .skipped:
+            DispatchQueue.main.async { [weak self] in
+                self?.customControls.learnMoreButton.isHidden = true
+                self?.customControls.progressBar.tintColor = .red
+                self?.customControls.progressBar.isUserInteractionEnabled = true
+                self?.customControls.skipButton.isHidden = true
+            }
+        default:
+            return
+        }
+    }
+
+    // MARK: - 5. Time observation for content
+
+    override func onMediaTimeEvent(_ time: JWTimeData) {
+        super.onMediaTimeEvent(time)
+        if progressBarScrubbing == false {
+            DispatchQueue.main.async { [weak self] in
+                self?.customControls.progressBar.maximumValue = Float(time.duration)
+                self?.customControls.progressBar.value = Float(time.position)
+            }
+        }
+    }
+
+
+    // MARK: - 6. Time observation for ads
+
+    override func onAdTimeEvent(_ time: JWTimeData) {
+        super.onAdTimeEvent(time)
+        DispatchQueue.main.async { [weak self] in
+            self?.customControls.progressBar.maximumValue = Float(time.duration)
```

Comment:
More commenting in here would be good, since this area of code is a little more complex.